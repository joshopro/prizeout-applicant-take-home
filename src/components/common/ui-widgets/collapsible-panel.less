.collapsible-panel {
    border: 2px solid #E5E7EA;
    border-radius: .5rem;
    position: relative;
    width: 100%;

    &.open {
        article {
            display: block;
        }

        header {
            &::after {
                content: '-' !important;    
            }

            border-bottom: 1px solid #E5E7EA;
            border-radius: .5rem;
        }
    }

    article,
    header {
        padding: .5rem;
    }

    article {
        color: #808997;
        display: none;
        font-size: .6rem;
        p {
            color: #808997;
            font-size: .6rem;
        }

        a {
            color: #00A7FF;
            text-decoration: none;
        }
    }

    header {
        &::after {
            content: '+';
            font-size: 1rem;
            line-height: 30px;
            position: absolute;
            right: 1rem;
            top: 0;
        }

        color: #808997;
        cursor: pointer;
    }
}

