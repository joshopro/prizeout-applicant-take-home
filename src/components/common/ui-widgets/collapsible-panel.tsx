import React, { useState } from 'react';
import Classnames from 'classnames';
import PropTypes from 'prop-types';
import { uuid } from 'uuidv4';

import './collapsible-panel.less';

interface CollapsiblePanelProps {
    content: string;
    id?: string;
    title: string;
}

export const CollapsiblePanel: React.FC<CollapsiblePanelProps> = ({ content, id, title }): React.ReactElement => {
    const [isOpen, setIsOpen] = useState(false),
        togglePanel = (): void => setIsOpen(!isOpen),
        classes: string = Classnames('collapsible-panel', { open: isOpen }),
        elmId = id || `cpc-${uuid()}`;

    return (
        <div className={classes}>
            <header aria-controls={elmId} aria-expanded={isOpen ? 'true' : 'false'} onClick={togglePanel} tabIndex={0}>
                {title}
            </header>
            <article dangerouslySetInnerHTML={{ __html: content }} id={elmId}></article>
        </div>
    );
};

CollapsiblePanel.propTypes = {
    content: PropTypes.string.isRequired,
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
};
