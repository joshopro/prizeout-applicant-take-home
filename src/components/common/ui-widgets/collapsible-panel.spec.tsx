import { render, screen } from '../../../testing/test-utils';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom';

import { CollapsiblePanel } from './collapsible-panel';

describe('Test Collapsible Panel Component', () => {
    const content = `Collapsible Panel Content`;
    const title = `Collapsible Panel Title`;

    test('Component matches snapshot', () => {
        const { asFragment } = render(<CollapsiblePanel content={content} title={title} id="test-cp" />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('Content is hidden by default', () => {
        render(<CollapsiblePanel content={content} title={title} />);

        // expect(screen.getByText(content)).not.toBeVisible();
    });

    test('Content shows when clicking on the header', () => {
        render(<CollapsiblePanel content={content} title={title} />);

        // expect(screen.getByText(content)).not.toBeVisible();
        userEvent.click(screen.getByText(title), { button: 0 });
        expect(screen.getByText(content)).toBeVisible();
    });

    test('Content hides when clicking on the header if it is showing', () => {
        render(<CollapsiblePanel content={content} title={title} />);

        // expect(screen.getByText(content)).not.toBeVisible();
        userEvent.click(screen.getByText(title), { button: 0 });
        expect(screen.getByText(content)).toBeVisible();
        userEvent.click(screen.getByText(title), { button: 0 });
        // expect(screen.getByText(content)).not.toBeVisible();
    });
});
