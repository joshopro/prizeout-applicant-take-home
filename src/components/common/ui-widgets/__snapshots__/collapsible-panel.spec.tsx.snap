// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Collapsible Panel Component Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="collapsible-panel"
  >
    <header
      aria-controls="test-cp"
      aria-expanded="false"
      tabindex="0"
    >
      Collapsible Panel Title
    </header>
    <article
      id="test-cp"
    >
      Collapsible Panel Content
    </article>
  </div>
</DocumentFragment>
`;
